pkg/epp/backend/metrics.go

```diff
@@ -0,0 +1,287 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backend
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/multierr"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datastore"
+	logutil "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/logging"
+)
+
+const (
+	// Hardcoded vLLM specific LoRA metrics
+	LoraRequestInfoRunningAdaptersMetricName = "running_lora_adapters"
+	LoraRequestInfoWaitingAdaptersMetricName = "waiting_lora_adapters"
+	LoraRequestInfoMaxAdaptersMetricName     = "max_lora"
+)
+
+type PodMetricsClientImpl struct {
+	MetricMapping *MetricMapping
+}
+
+// FetchMetrics fetches metrics from a given pod.
+func (p *PodMetricsClientImpl) FetchMetrics(
+	ctx context.Context,
+	existing *datastore.PodMetrics,
+	port int32,
+) (*datastore.PodMetrics, error) {
+	logger := log.FromContext(ctx)
+	loggerDefault := logger.V(logutil.DEFAULT)
+
+	url := "http://" + existing.Address + ":" + strconv.Itoa(int(port)) + "/metrics"
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		loggerDefault.Error(err, "Failed create HTTP request", "method", http.MethodGet, "url", url)
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		loggerDefault.Error(err, "Failed to fetch metrics", "pod", existing.NamespacedName)
+		return nil, fmt.Errorf("failed to fetch metrics from %s: %w", existing.NamespacedName, err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		loggerDefault.Error(nil, "Unexpected status code returned", "pod", existing.NamespacedName, "statusCode", resp.StatusCode)
+		return nil, fmt.Errorf("unexpected status code from %s: %v", existing.NamespacedName, resp.StatusCode)
+	}
+
+	parser := expfmt.TextParser{}
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return p.promToPodMetrics(logger, metricFamilies, existing)
+}
+
+// promToPodMetrics updates internal pod metrics with scraped Prometheus metrics.
+func (p *PodMetricsClientImpl) promToPodMetrics(
+	logger logr.Logger,
+	metricFamilies map[string]*dto.MetricFamily,
+	existing *datastore.PodMetrics,
+) (*datastore.PodMetrics, error) {
+	var errs error
+	updated := existing.Clone()
```

Comment:
@liu-cong perhaps we can address this in your PR, but why are we cloning here? Just create a Metrics (not PodMetrics) struct and return it.

Reply:
yeah this is addressed in https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/460/files